services/fission-rmq-connector/main.go

```diff
@@ -0,0 +1,219 @@
+	if err != nil {
+		return err
+	}
+
+	log.Printf("Consuming from queue: %s and producing to exchange: %s\n", AMQP_QUEUE, AMQP_XCHG)
+	conn := mq.NewConnection(AMQP_HOST)
+	successChan := conn.Publisher(AMQP_XCHG, func(c *amqp091.Channel) error {
+		return nil
+	})
+	consumeChan := conn.Consume(AMQP_QUEUE, func(c *amqp091.Channel) error {
+		// Cannot set autodelete because KEDA can't scale if the queue does not exist at all
+		_, err := c.QueueDeclare(AMQP_QUEUE, true, false, false, false, nil)
+		if err != nil {
+			return err
+		}
+		err = c.ExchangeDeclare(AMQP_XCHG, "topic", true, false, false, false, nil)
+		if err != nil {
+			return err
+		}
+		err = c.QueueBind(AMQP_QUEUE, AMQP_QUEUE, AMQP_XCHG, false, nil)
+		if err != nil {
+			return err
+		}
+		err = c.Qos(prefetch, 0, true)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	go conn.Start()
+
+	connector := Connector{
+		Name:       fmt.Sprintf("%s-(%s)", os.Getenv("SOURCE_NAME"), os.Getenv("HOSTNAME")),
+		Endpoint:   HTTP_ENDPOINT,
+		MaxRetries: maxRetries,
+		Result:     successChan,
+	}
+
+	concurrency := make(chan int, prefetch)
+	go func(incoming <-chan amqp091.Delivery) {
+		for delivery := range incoming {
+			concurrency <- 1
+			go func(delivery amqp091.Delivery) {
+				log.Printf("Processing: %s\n", delivery.MessageId)
+				connector.handleDelivery(delivery)
+				<-concurrency
+			}(delivery)
+		}
+	}(consumeChan)
+
+	if METRICS_ADDR != "" {
+		go func() {
+			srv := &http.Server{
+				Addr:         METRICS_ADDR,
+				WriteTimeout: 5 * time.Second,
+				ReadTimeout:  5 * time.Second,
+				Handler:      promhttp.Handler(),
+			}
+			log.Printf("Metrics server starting at: %s\n", METRICS_ADDR)
+			err := srv.ListenAndServe()
+			log.Printf("Metrics server stopped: %s\n", err.Error())
+		}()
+	}
+
+	log.Printf("RabbitMQ-Fission Connector is running...\n")
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
+	defer cancel()
+	<-ctx.Done()
+	log.Printf("RabbitMQ-Fission interrupted, shutting down gracefully...\n")
+
+	conn.Shutdown()
+
+	return nil
+}
+
+type Connector struct {
+	Name       string
+	Endpoint   string
+	MaxRetries int
+	Result     chan<- mq.PublishMessage
+}
+
+func (c *Connector) handleDelivery(delivery amqp091.Delivery) {
+	opsProcessed.Inc()
+
+	res, err := doHTTPRequest(delivery.Body, c.Endpoint, c.MaxRetries)
+	if err != nil {
+		// This is a Function Invocation error or a fatal worker error
+		// can't be a business logic error (ie device not found) such an error would be considered
+		// a succesful invocation
+		opsFails.Inc()
+		c.handleError(delivery, err)
+	} else {
+		opsSuccesful.Inc()
+		c.handleSuccess(delivery, res)
+	}
+}
+
+func (c *Connector) handleError(delivery amqp091.Delivery, err error) {
+	// The invocation failed, what to do?
+	log.Printf("Invocation error: %v. Redelivering?: %v", err.Error(), !delivery.Redelivered)
+	if !delivery.Redelivered {
+		delivery.Nack(false, true)
```

Comment:
is it automatically acked when it has been redelivered?

Reply:
Yes, this is to ensure that a message is not infinitely looped in the queue.